tortoise/rerun.go

```diff
@@ -116,7 +118,7 @@ type validityTracer struct {
 func (vt *validityTracer) SaveContextualValidity(bid types.BlockID, lid types.LayerID, validityNew bool) error {
 	if vt.firstUpdatedLayer == nil {
 		validityCur, err := vt.ContextualValidity(bid)
-		if err != nil {
+		if err != nil && !errors.Is(err, database.ErrNotFound) {
```

Comment:
this is tricky....
this only make sense if `lid` is later than the verified layer. but if `lid` is older than verified layer, all block contextual validity should be saved already. so database.ErrNotFound should be programming error and should abort rerun as well.

this led me to think, should we only do rerun up to currently verified layer? why should rerun compete with regular tortoise?

Reply:
> this only make sense if lid is later than the verified layer. but if lid is older than verified layer, all block contextual validity should be saved already. so database.ErrNotFound should be programming error and should abort rerun as well.

blocks can be missing when tortoise verified layer initially

> this led me to think, should we only do rerun up to currently verified layer? why should rerun compete with regular tortoise?

we need to rerun until the last layer anyway to have compatible state, so i am not sure that i can make this change

Reply:
> blocks can be missing when tortoise verified layer initially. because of this problem the rerun was failing on miner-4
right. but if `database.ErrNotFound` were to happen to an already verified layer, then it's a honest programming error.

> we need to rerun until the last layer anyway to have compatible state, so i am not sure that i can make this change
not asking you to change in this PR. but i am questioning the rerun design. the point of rerun is to double-check if we are still confident of what we've done so far. as i understand it, this need arose when we changed from vote-counting tortoise to verifying tortoise. i think we should take it up to Tal again for the more thorough discussion.

Reply:
> right. but if database.ErrNotFound were to happen to an already verified layer, then it's a honest programming error.

i don't think so, in case of self-healing after long lasting partition layer will be verified but every side will be missing some blocks

Reply:
> > right. but if database.ErrNotFound were to happen to an already verified layer, then it's a honest programming error.
> 
> i don't think so, in case of self-healing after long lasting partition layer will be verified but every side will be missing some blocks

i agree. i meant programming error in the sense where we previously find this block contextually valid but didn't save it correctly. in any case, i think i'm splitting hair here. the point of rerun is to fix what was missing previously. so i am agreeing with you now.